### PR TITLE
Integrate `ptxas` compiler to `iree-compile`

### PR DESCRIPTION
**Pull request by grypp**
`iree-compile` generates PTX and puts into vmfb file. The PTX is compiled on the runtime using `cuModuleLoadDataEx`. This is insufficient in three ways:
- Cannot try different `ptxas` compiler version.
- `ptxas` might behave differently than `cuModuleLoadDataEx`.
- Keeping PTX increases vmfb file size, but also great for jit compile. 

The PR implements followings:
- Introduces three flags in the `iree-compile` : `--iree-hal-use-ptxas=true/false`, `--iree-hal-use-ptxas-from=<path>`, `--iree-hal-use-ptxas-params=<params>`
- When `--iree-hal-use-ptxas=true`, compiles the iree generated PTX and keeps it as `cubin` in the `vmfb` file. This reduces the file size significantly.
- If `--iree-hal-use-ptxas-from=<path>` is not present, it searches the `ptxas` from the path. 
- When `--iree-hal-use-ptxas=false`, packs the ptx into vmfb file, and let the runtime to compile.

Flags can use them like below :
```
iree-compile code.mlir --iree-hal-use-ptxas-from=/usr/local/cuda-11.8/bin/ptxas -o code.vmfb
NOTE: Compiling the generated PTX code
 $ /usr/local/cuda-11.8/bin/ptxas -arch sm_80  /tmp/iree-cuda-ptx-src-1a146b -o /tmp/iree-cuda-ptx-src-1a146b.cubin 2> /tmp/iree-cuda-ptx-log-0de384
```

One can pass `-v` to see extra information such as register spilling, static shared and local memory usage and etc.
```
iree-compile code.mlir --iree-hal-use-ptxas-from=/usr/local/cuda-11.8/bin/ptxas --iree-hal-use-ptxas-params=-v -o code.vmfb

NOTE: Compiling the generated PTX code
 $ /usr/local/cuda-11.8/bin/ptxas -arch sm_80 -v /tmp/iree-cuda-ptx-src-ceb4df -o /tmp/iree-cuda-ptx-src-ceb4df.cubin 2> /tmp/iree-cuda-ptx-log-dec4da
ptxas info    : 0 bytes gmem
ptxas info    : Compiling entry function 'matmul_dispatch_0_matmul_1024x1024x1024' for 'sm_80'
ptxas info    : Function properties for matmul_dispatch_0_matmul_1024x1024x1024
    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 188 registers, 376 bytes cmem[0]
```

### CODE REVIEW

**Comment by grypp** [comment]
Let me explain in detail why do we want `ptxas` instead `cuModuleLoadDataEx` or even calling different versions of `ptxas`. 

```
#guray = #iree_codegen.compilation_info<
  lowering_config = <tile_sizes = [[128, 128, 32]]>,
  translation_info = <LLVMGPUMatmulTensorCore
  pipeline_depth = 3>,
  workgroup_size = [128 : index, 1 : index, 1 : index]
> 
func.func @matmul(%lhs: tensor<6912x6912xf32>,  %rhs: tensor<6912x6912xf32>,
  %inital_result: tensor<6912x6912xf32>) -> tensor<6912x6912xf32> {
  %cst = arith.constant 1.0 : f32
  %out = linalg.fill ins(%cst : f32) outs(%inital_result : tensor<6912x6912xf32>) -> tensor<6912x6912xf32>
  %result = linalg.matmul {compilation_info = #guray} 
                     ins(%lhs, %rhs: tensor<6912x6912xf32>, tensor<6912x6912xf32>)
                     outs(%out: tensor<6912x6912xf32>) -> tensor<6912x6912xf32>
  return %result : tensor<6912x6912xf32>
}
```

I took the code above and compiled 6 versions. Note that, `iree-flow-add-cutlass` add `cutlass` to the CUDA kernel name. 

```
set -gx FLAGS --iree-hal-benchmark-dispatch-repeat-count=5 --iree-hal-target-backends=cuda --iree-hal-cuda-llvm-target-arch=sm_80
iree-compile $FLAGS test.mlir --iree-flow-add-cutlass=false -o test-driver.vmfb         -iree-codegen-llvmgpu-use-mma-sync  &
iree-compile $FLAGS test.mlir --iree-flow-add-cutlass=false -o test-cuda118.vmfb        -iree-codegen-llvmgpu-use-mma-sync --iree-hal-use-ptxas-from=/usr/local/cuda-11.8/bin/ptxas    &
iree-compile $FLAGS test.mlir --iree-flow-add-cutlass=false -o test-cuda116.vmfb        -iree-codegen-llvmgpu-use-mma-sync --iree-hal-use-ptxas-from=/usr/local/cuda-11.6/bin/ptxas    &
iree-compile $FLAGS test.mlir --iree-flow-add-cutlass=true  -o test-cutlass-driver.vmfb -iree-codegen-llvmgpu-use-mma-sync   &
iree-compile $FLAGS test.mlir --iree-flow-add-cutlass=true  -o test-cutlass-cuda118.vmfb -iree-codegen-llvmgpu-use-mma-sync --iree-hal-use-ptxas-from=/usr/local/cuda-11.8/bin/ptxas &
iree-compile $FLAGS test.mlir --iree-flow-add-cutlass=true  -o test-cutlass-cuda116.vmfb -iree-codegen-llvmgpu-use-mma-sync --iree-hal-use-ptxas-from=/usr/local/cuda-11.6/bin/ptxas &
```

`test-cutlass-driver.vmfb` and `test-driver.vmfb` keep PTX and they are significantly larger files. Others keeps the cubin as binary. 
```
^^/r/slowPtx >>> ll *.vmfb
-rw-rw-r-- 1 gurayozen gurayozen 22K Mar 29 10:27 test-cuda116.vmfb
-rw-rw-r-- 1 gurayozen gurayozen 22K Mar 29 10:27 test-cuda118.vmfb
-rw-rw-r-- 1 gurayozen gurayozen 22K Mar 29 10:27 test-cutlass-cuda116.vmfb
-rw-rw-r-- 1 gurayozen gurayozen 22K Mar 29 10:27 test-cutlass-cuda118.vmfb
-rw-rw-r-- 1 gurayozen gurayozen 62K Mar 29 10:27 test-cutlass-driver.vmfb
-rw-rw-r-- 1 gurayozen gurayozen 62K Mar 29 10:27 test-driver.vmfb
```

I run these vmfb files using ncu profiler, here are the result. I have cuda 11.8 driver. But there is substantial difference between `test-cutlass-driver.vmfb` vs `test-cutlass-cuda116.vmfb`. Having `ptxas` flags in hand helps to a lot here to identify the performance regression. 

```
Executable                   Elapsed Cycles 
test-driver.vmfb            ->  9396807
test-cuda118.vmfb           ->  9385476
test--cuda116.vmfb          ->  9391230
test-cutlass-driver.vmfb    ->  8565386
test-cutlass-cuda118.vmfb   ->  8520628
test-cutlass-cuda116.vmfb   ->  9388590
```

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (qcolombet) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: Use `StringRef` instead of `std::string` for the arguments.

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (qcolombet) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: What does the absolute path requirement gives us?

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (qcolombet) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: Could be a follow-up PR, but I would love an option to keep the intermediate files just to double check that the right source ptx has been used when we start mixing things with `--iree-hal-substitute-executable-object` for instance.

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (qcolombet) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: You may be able to use Twine here as well to avoid the temporary object.

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (qcolombet) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: Shouldn't the file remover occur earlier (line around line 90)?

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (qcolombet) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: Could you use the actual type instead of `auto`?
It's unclear to me what we get here.

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (qcolombet) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: I don't think we need the temporay `std::string` here.

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (qcolombet) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: Use the actual type instead of `auto`

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (qcolombet) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: I don't know what's the general rule here, if any, but a lot of options in IREE starts with `cl`. I.e., here that would be  `clUsePTASfrom`.
I'm fine either way.

**Review by qcolombet — APPROVED** [review]
I'd like some people more familiar with IREE to comment here, but this LGTM. (I.e., probably wait for another reviewer before landing anything.)

I tried it locally and I found it super useful to be able to change which cuda toolkit is used (for the compilation) on the fly.

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (grypp) [reply compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: Good point. I tried that, it works with `--iree-hal-substitute-executable-object`. You can compile your ptx with ptxas compiler like below
```
iree-compile $FLAGS test.mlir -o test-116.vmfb \
--iree-hal-use-ptxas-from=/usr/local/cuda-11.6/bin/ptxas \
--iree-hal-use-ptxas-params=-v \
--iree-hal-substitute-executable-object=matmul_dispatch_0=my_fast.ptx
```

```
NOTE: hal.executable `matmul_dispatch_0` substituted with object file at `my_fast.ptx`
NOTE: Compiling the generated PTX code 
 $ /usr/local/cuda-11.6/bin/ptxas -arch sm_80 -v /tmp/iree-cuda-ptx-src-9713c0 -o /tmp/iree-cuda-ptx-src-9713c0.cubin 2> /tmp/iree-cuda-ptx-log-139a37
ptxas info    : 0 bytes gmem
ptxas info    : Compiling entry function 'matmul_dispatch_0_matmul_2048x2048x2048' for 'sm_80'
ptxas info    : Function properties for matmul_dispatch_0_matmul_2048x2048x2048
    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 189 registers, 376 bytes cmem[0]
```

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: this is not ptx anymore here?

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: how does the runtime know whether it is getting ptx or cubin? It needs to call a different API right?

Comment (grypp) [reply compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: IREE runtime doesn't need to know. We call the same function `cuModuleLoadEx`, but pass the `cubin` that is specifically compiled for the target GPU. `cuModuleLoadEx` can directly use the `.cubin` if the target GPU is same as the compiled one.

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: ah cool, I didn't know

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: please use LLVM variable naming style here and below.

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (grypp) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: Is it good way to find default `ptxas`?

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (grypp) [reply compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: no it's actually cubin, let me change the name.

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (grypp) [reply compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: By the way we can also embed the ptx in vmfb as well, so if the binary in `cubin` is not suitable for the target GPU, we can jit compile.

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp]: > By the way we can also embed the ptx in vmfb as well, so if the binary in `cubin` is not suitable for the target GPU, we can jit compile.

yeah let's keep it simple for now

**Comment by powderluv** [comment]
Does this also work on windows ?

**Comment by grypp** [comment]
> Does this also work on windows ?

I don't have windows here to try it out. I used llvm's api for file management and invoking the `ptxas`. It might just work

**Review by ThomasRaoux — APPROVED** [review]
This is great, should be very useful.

**Comment by manishucsd** [comment]
Thanks for this analysis and adding the PR. Focusing on the below runs which has the same cuda/driver version of 11.8. Please confirm that test-driver.vmfb  and test-cutlass-driver.vmfb are using`520.61.05` driver installed as `sudo sh cuda_11.8.0_520.61.05_linux.run --driver` ([see link](https://developer.nvidia.com/cuda-11-8-0-download-archive?target_os=Linux&target_arch=x86_64&Distribution=Ubuntu&target_version=22.04&target_type=runfile_local))

Based on the above:
```bash
test-driver.vmfb            ->  9396807
test-cutlass-driver.vmfb    ->  8565386
test-cuda118.vmfb           ->  9385476
test-cutlass-cuda118.vmfb   ->  8520628
```

Normalizing it with the fastest run (Numbers show slowdown w.r.t. ptx compiled using ptxas from cuda 11.8 toolkit ):
```bash
test-driver.vmfb            ->  1.10
test-cutlass-driver.vmfb    ->  1.005
test-cuda118.vmfb           ->  1.10
test-cutlass-cuda118.vmfb   ->  1.00
```

Both driver's JIT compiled ptx-to-sass and toolkit's ptxas compilation to sass reaches the fastest performance with `--iree-flow-add-cutlass=true`. Is this a correct assertion?

**Review by manishucsd — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp
Comment (manishucsd) [inline compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp:66]: `--iree-hal-use-ptxas` "It uses the default ptxas compiler." 
Is default ptxas compiler means using driver's JIT ptx-to-saas flow? What does default mean here?

From the PR description:
```
When --iree-hal-use-ptxas=false, packs the ptx into vmfb file, and let the runtime to compile.
Set--iree-hal-use-ptxas to true. So default compilation is ptxas.
```

it seems like `--iree-hal-use-ptxas=false` is the current behaviour of iree-compile, i.e., ptx in vmfb file is JIT compiled to SASS by the driver. 

`--iree-hal-use-ptxas=true` uses ptxas provided by `--iree-hal-cuda-use-ptxas-from`

Should the default setting for --iree-hal-use-ptxas be false setting it to the status quo? Let me know if I am missing something here.

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp:66]: yes should be changed to `false`

Comment (grypp) [reply compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp:66]: I clarified the text here, `--iree-hal-use-ptxas` finds `ptxas` on your PATH. 
I changed the default to `--iree-hal-use-ptxas=false`. 

as a future work, we could make default compilation `cubin` + `ptx`. If `cubin` is not for the target, we jit `ptx` on the runtime.

**Comment by grypp** [comment]
> Thanks for this analysis and adding the PR. Focusing on the below runs which has the same cuda/driver version of 11.8. Please confirm that test-driver.vmfb and test-cutlass-driver.vmfb are using`520.61.05` driver installed as `sudo sh cuda_11.8.0_520.61.05_linux.run --driver` ([see link](https://developer.nvidia.com/cuda-11-8-0-download-archive?target_os=Linux&target_arch=x86_64&Distribution=Ubuntu&target_version=22.04&target_type=runfile_local))
> 
> Based on the above:
> 
> ```shell
> test-driver.vmfb            ->  9396807
> test-cutlass-driver.vmfb    ->  8565386
> test-cuda118.vmfb           ->  9385476
> test-cutlass-cuda118.vmfb   ->  8520628
> ```
> 
> Normalizing it with the fastest run (Numbers show slowdown w.r.t. ptx compiled using ptxas from cuda 11.8 toolkit ):
> 
> ```shell
> test-driver.vmfb            ->  1.10
> test-cutlass-driver.vmfb    ->  1.005
> test-cuda118.vmfb           ->  1.10
> test-cutlass-cuda118.vmfb   ->  1.00
> ```
> 
> Both driver's JIT compiled ptx-to-sass and toolkit's ptxas compilation to sass reaches the fastest performance with `--iree-flow-add-cutlass=true`. Is this a correct assertion?

There are cases where the driver's JIT compiles to ptx-to-sass and the toolkit's ptxas build produces different SASS even if the versions are the same. Also, I can't say that a version of ptx compiler is fast for any program. This depends on your existing PTX. This was my past experience as well.

That's why having chance to choose different versions of driver and `ptxas` compilers is useful, so we can see where do we loose/gain performance.

**Pull request merged**
